cargo-miri/bin.rs

```diff
@@ -249,27 +249,57 @@ fn xargo_check() -> Command {
     Command::new(env::var_os("XARGO_CHECK").unwrap_or_else(|| OsString::from("xargo-check")))
 }
 
-/// Execute the command. If it fails, fail this process with the same exit code.
-/// Otherwise, continue.
-fn exec(mut cmd: Command) {
-    let exit_status = cmd.status().expect("failed to run command");
-    if exit_status.success().not() {
+/// Execute the `Command`, where possible by replacing the current process with a new process
+/// described by the `Command`. Then exit this process with the exit code of the new process.
+fn exec(mut cmd: Command) -> ! {
+    // On non-Unix imitate POSIX exec as closely as we can
+    #[cfg(not(unix))]
+    {
+        let exit_status = cmd.status().expect("failed to run command");
         std::process::exit(exit_status.code().unwrap_or(-1))
     }
+    // On Unix targets, actually exec
+    // If exec returns, process setup has failed. This is the same error condition as the expect in
+    // the non-Unix case.
+    #[cfg(unix)]
+    {
+        use std::os::unix::process::CommandExt;
+        let error = cmd.exec();
+        Err(error).expect("failed to run command")
+    }
 }
 
-/// Execute the command and pipe `input` into its stdin.
-/// If it fails, fail this process with the same exit code.
-/// Otherwise, continue.
-fn exec_with_pipe(mut cmd: Command, input: &[u8]) {
-    cmd.stdin(process::Stdio::piped());
-    let mut child = cmd.spawn().expect("failed to spawn process");
+/// Execute the `Command`, where possible by replacing the current process with a new process
+/// described by the `Command`. Then exit this process with the exit code of the new process.
+/// `input` is also piped to the new process's stdin, on cfg(unix) platforms by writing its
+/// contents to `path` first, then setting stdin to that file.
+fn exec_with_pipe<P>(mut cmd: Command, input: &[u8], path: P) -> !
+where
+    P: AsRef<Path>,
+{
+    #[cfg(unix)]
     {
-        let stdin = child.stdin.as_mut().expect("failed to open stdin");
-        stdin.write_all(input).expect("failed to write out test source");
+        // Write the bytes we want to send to stdin out to a file
+        std::fs::write(&path, input).unwrap();
+
+        // Open the file for reading, and set our new stdin to it
+        let stdin = File::open(&path).unwrap();
+        cmd.stdin(stdin);
+
+        // Unlink the file so that it is fully cleaned up as soon as the new process exits
+        std::fs::remove_file(&path).unwrap();
+        exec(cmd)
```

Comment:
```suggestion
        exec(cmd)
```
Why is this in the code paragraph titles "unlink the file"?
I don't quite understand how you are using empty lines here.

Reply:
To be honest I don't always understand them either. I dropped the empty lines for now, but I could totally put some back if you think they help. I know a lot of the rust-lang code doesn't use them, but I'm used to it. Just a personal thing.

Reply:
I also use them, basically like paragraphs -- blocks of code without empty lines between them should form "logical units" or so.